### ISSUE TITLE
inject license info into asset metadata on save/publish

### ISSUE THREAD

**Issue by satra** [dandi/dandi-api#461]
currently license is only at the dandiset level. as part of our metadata injection process (for non-published assets), we should add the dandiset license into the asset metadata.  to prevent doing a migration again, we can check for this during publish and inject the license info. 

published assets license cannot change. 

presently the schema does not require this for an asset, but i think we should do this for appropriate licensing (especially for future use in other dandisets).

**Comment by satra** [comment]
@yarikoptic - cc:ing here as well.

**Comment by yarikoptic** [comment]
Not sure why act of publishing is the one to inject license for assets - license might have been known since beginning.

Do we have copyright field in metadata? (License without a copyright - ownership- is kinda like a driver license without a driver AFAIK)

Not necessarily a single asset would have a license (factual data points aren't copyrightable) while dataset might as representing a unique collection of data

I see why we might like to have explicit license within reach asset, but I think we might not want to assume that such assignment would be intended by the user. 

Again - the same draft asset might be shared by multiple dandisets. Winner to publish it first mutates it. Injecting more of that particular dandiset metadata would add to the confusion IMHO

**Comment by satra** [comment]
just like license added to dandiset metadata, license is added to asset metadata (there is no conceptual difference as far as i can tell). as a piece of structured text it has implicit copyright. the licenses we are using apply to data as well. i can dig up the reference if needed, but there is a discussion on data licenses, and since i'm not a lawyer, i will take the statements for granted that legal experts have looked at it.

regarding the mutation of licenses. this is the technical thing we need to implement:

1. a draft asset cannot be added to another dataset unless the person adding is one of the owners of the asset
2. a published asset's license never changes. (this is a bit more tricky but doable - since we currently don't have a way of keeping asset id intact if path mutates in a secondary dataset, we would have to rely on something like wasDerivedFrom - another asset, and that assets license would carry through ).

regarding licenses, currently the only place that allows easy entry is the dandiset metadata. the cli could also inject on upload. but we need to think through the use cases of subsetting (use a subset of assets), aggregate (use a collection of assets from different sources), modification (add to the data blob itself). all of this is really about provenance, which is where asset ids become important. 

it's just that we can make life a little easier than simply leaving it to the users, who most of the time don't think about this.

**Comment by waxlamp** [comment]
To test my understanding, I want to summarize the proposed policy in different words and see if it rings true to you:
1. Assets are assigned the license of their dandiset upon publish.
2. A draft asset can only be added to a different dandiset by the owner of the dandiset containing that asset.
3. A published asset can be added to a dandiset if the asset's license is compatible with the dandiset's license.

Is this right?

I am having some trouble with my point (2) above. To take an extreme hypothetical example: suppose I have draft dandiset A with draft asset B, and I intend to publish A with a permissive license of some kind. Now suppose I create dandiset C, and I add asset B to it (which I can do because I own A and C both). Now I publish A, then I set a restrictive license on C. What happens when I attempt to publish C? I assume the validation step would tell me that the licensing doesn't match and I can't publish in the current state. But in this case, why make restrictions such as (2)?

It feels like perhaps any asset should be addable to any dandiset, but that in many such cases there would be an immediate error report about incompatibility of licenses, etc.

Or perhaps this is moot because we are not allowing any incompatible licenses in the system?

**Comment by satra** [comment]
@waxlamp - yes on first pass your thought process is correct. 

the key is to think about provenance of the asset, and while we think of asset as metadata + blob, at least presently the license tied to this joint construct, but provenance and hence license could be tracked from the perspective of the earliest point. to make things more complicated:
1. licenses can change on drafts
2. any metadata change right now mints new asset ids, so a simple path change can modify linkage.

i do think we need our graph database with provenance tracking of drafts in place before we can truly consider the issues at play. but let's continues discussing the rationale and the possible ramifications of this potential change.

**Comment by waxlamp** [comment]
I think in the short term, all we really need is to validate license compatibility before publish. (To see why this is true, I could rephrase this in a shifted context as "all we really need is to validate *existence* of a license before publish", which is something we already do, or "all we really need is to validate that at least one asset exists before publish", etc.)

But this brings up a new issue, specific to licenses: if I publish an asset under license A, and then someone puts that asset in their dandiset, our system currently allows them to create a new asset simply by changing the license on their "copy" of the asset. But a license, by definition, isn't something that anyone can change, so this seems to expose a weakness in our semantics of assets and licensing.

By contrast, we *don't* have this problem with dandisets as a whole. Only an owner of a dandiset can change the license, and we don't have any built-in primitives for copying a dandiset wholesale, whereby some new owner could obtain a copy of a dandiset and then (illegally) change the license, etc. (They could, by contrast, recreate an existing dandiset beat-for-beat by hand, except they would be the owner of it, but that's the same problem as anyone being able to self-publish a word-for-word copy of _Great Expectations_, for example. That is, it's a way in which a person can achieve an "illegal" outcome but it is outside the scope of what the system can physically restrain.)

I may still be missing out on some subtleties of the intended meaning of "ownership" and "licensing" in this context. But I do know that scientists are (naturally) jealous about their data so it is probably a good idea to sort out our overall goal with asset and dandiset licensing semantics.

**Comment by waxlamp** [comment]
We probably need some more discussion on this. Re-reading this thread, I'm still not super clear on what problem this is trying to solve, nor where this leaves all the stickier concepts around the semantics of licensing assets and dandisets together.
